### PR TITLE
Complete ESP32 PoC for ADU OTA 

### PR DESCRIPTION
**Pull request by ewertons**
## Purpose

Celebrate the union of Azure SDK for C, Middleware for FreeRTOS and ADU.

## Does this introduce a breaking change?

```
[x] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

### CODE REVIEW

File: demos/projects/ESPRESSIF/aduotaupdate/partitions_ota.csv [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Name,   Type, SubType, Offset,   Size, Flags
 # Note: if you have increased the bootloader size, make sure to update the offsets to avoid overlap
-nvs,      data, nvs,     ,        0x4000,
+nvs,      data, nvs,     ,        0x6000,
```
Comment (danewalton) [inline demos/projects/ESPRESSIF/aduotaupdate/partitions_ota.csv:3]: what was this expanded for?

File: demos/projects/ESPRESSIF/aduotaupdate/port/azure_iot_flash_platform.c [hunk 2]
```diff
@@ -10,12 +10,25 @@
 
 AzureIoTResult_t AzureIoTPlatform_Init( AzureADUImage_t * const pxAduImage )
 {
+  const esp_partition_t * pxCurrentPartition = esp_ota_get_running_partition();
+
+  if ( pxCurrentPartition == NULL )
+  {
+    printf( ( "esp_ota_get_running_partition failed" ) );
+    return eAzureIoTErrorFailed;
+  }
```
Comment (danewalton) [inline demos/projects/ESPRESSIF/aduotaupdate/port/azure_iot_flash_platform.c:19]: good change :)

File: demos/sample_azure_iot_adu/sample_azure_iot_adu.c [hunk 3]
```diff
@@ -155,9 +156,17 @@ static uint8_t ucCommandResponsePayloadBuffer[ 256 ];
 static uint8_t ucReportedPropertiesUpdate[ 1500 ];
 static uint32_t ulReportedPropertiesUpdateLength;
 
-static uint8_t ucAduContextBuffer[ 1024 ];
+static uint8_t ucAduContextBuffer[ 10240 ];
```
Comment (danewalton) [inline demos/sample_azure_iot_adu/sample_azure_iot_adu.c:159]: ooof that's a big buffer. it is what it is :)

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c [hunk 4]
```diff
@@ -350,7 +342,11 @@ AZ_NODISCARD az_result az_iot_adu_ota_parse_service_properties(
 
                     _az_RETURN_IF_FAILED(az_json_token_get_string(
                         &jr->token, (char*)az_span_ptr(update_request->workflow.id),
-                        az_span_size(update_request->workflow.id), NULL));
+                        az_span_size(update_request->workflow.id), &out_length));
+
+                    // TODO: find a way to get rid of az_json_token_get_string (which adds a \0 at the end!!!!!!)
+                    //       Preferably have a function that does not copy anything.
+                    update_request->workflow.id = az_span_slice(update_request->workflow.id, 0, out_length);
```
Comment (danewalton) [inline libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c:349]: is it really a problem that it appends a `NULL` character? Just out of curiosity. We could always create a span around the returned char* and do a `size - 1`. Or is it that doing that is clunky?

File: libs/azure-iot-middleware-freertos/ports/coreHTTP/azure_iot_core_http.c [hunk 5]
```diff
@@ -136,7 +136,7 @@ AzureIoTHTTPResult_t AzureIoTHTTP_Request( AzureIoTHTTPHandle_t xHTTPHandle,
         char headerbuffer[ 32 ] = { 0 };
         int valueLength = snprintf( headerbuffer, sizeof( headerbuffer ), "bytes=%d-%d", lRangeStart, lRangeEnd );
 
-        xHttpLibraryStatus = HTTPClient_AddHeader( &xHTTPHandle->xRequestHeaders, "x-ms-range", sizeof( "x-ms-range" ) - 1,
+        xHttpLibraryStatus = HTTPClient_AddHeader( &xHTTPHandle->xRequestHeaders, "Range", sizeof( "Range" ) - 1,
```
Comment (danewalton) [inline libs/azure-iot-middleware-freertos/ports/coreHTTP/azure_iot_core_http.c:139]: We can replace this with `HTTPClient_AddRangeHeader()`.

https://github.com/FreeRTOS/coreHTTP/blob/ff33d4645807d36e314ce074597a15b70782969e/source/include/core_http_client.h#L715-L718

Comment (danewalton) [reply libs/azure-iot-middleware-freertos/ports/coreHTTP/azure_iot_core_http.c:139]: No need for a temporary buffer

**Review by danewalton — APPROVED** [review]

File: demos/projects/ESPRESSIF/aduotaupdate/partitions_ota.csv (hunk 1 above)
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/partitions_ota.csv:3]: To allow the network stack to have enough memory.
I was getting this failure with 0x4000:
ESP_ERROR_CHECK failed: esp_err_t 0x110d (ESP_ERR_NVS_NO_FREE_PAGES) at 0x400870f0
when calling nvs_flash_init

File: demos/projects/ESPRESSIF/aduotaupdate/port/azure_iot_flash_platform.c (hunk 2 above)
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/port/azure_iot_flash_platform.c:19]: I had to, to debug some errors.

File: demos/sample_azure_iot_adu/sample_azure_iot_adu.c (hunk 3 above)
Comment (ewertons) [reply demos/sample_azure_iot_adu/sample_azure_iot_adu.c:159]: I was oooooverly lax with the sizes for now. Next we will need to trim all these buffers.

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c (hunk 4 above)
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c:349]: My plan is to change this function to not use the az_token_get_string, only assign the token slice (and avoid copies). So to maintain the same public contract of no extra '\0's, I trimmed them here.

File: libs/azure-iot-middleware-freertos/ports/coreHTTP/azure_iot_core_http.c (hunk 5 above)
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/ports/coreHTTP/azure_iot_core_http.c:139]: I'll make that change.

Comment (ewertons) [reply libs/azure-iot-middleware-freertos/ports/coreHTTP/azure_iot_core_http.c:139]: Done. And it works great. Thanks.

**Pull request merged**